### PR TITLE
First pass at adding MLSA support

### PR DESCRIPTION
**Pull request by tyler-ball**
Signed-off-by: tyler-ball <tball@chef.io>

## Description
We need to update the following products to prompt users for MLSA instead of EULA:

* Chef Workstation
* Chef DK
* Push Jobs Client
* Supermarket

The following products also need to be updated from EULA to MLSA but we need to think about how we want to do this. They are installed as plugins to Chef Server, which is covered under the EULA.
* Chef Manage
* Push Jobs Server

License acceptance was never added for the following, and will need to be implemented:
* Chef Backend

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by tas50** [comment]
Maybe this isn't the place for this, but WTF. Why don't we just use the new EULA on everything.

File: components/ruby/config/product_info.toml [hunk 1]
```diff
@@ -1,4 +1,3 @@
-# DO NOT CHANGE - overwritten by expeditor at build time
```
Comment (marcparadise) [inline components/ruby/config/product_info.toml:1]: This is no longer a true statement?

File: components/ruby/example/example_mixlib_cli_app.rb [hunk 2]
```diff
@@ -35,7 +35,7 @@ def run(argv = ARGV)
         options[:provided] = ENV["DEBUG_LICENSE"]
       end
 
-      LicenseAcceptance::Acceptor.check_and_persist!("infra-client", "1337", options)
+      LicenseAcceptance::Acceptor.check_and_persist!("chef-workstation", "1337", options)
```
Comment (marcparadise) [inline components/ruby/example/example_mixlib_cli_app.rb:38]: Nit: why the change?

File: components/ruby/lib/license_acceptance/strategy/file.rb [hunk 3]
```diff
@@ -29,11 +29,18 @@ def accepted?(product_relationship)
...
-            f = ::File.join(loc, product.filename)
-            if ::File.exist?(f)
-              logger.debug("Found license #{product.filename} at #{f}")
-              missing_licenses.delete(product)
-              break
+            license_path = ::File.join(loc, product.filename)
+            if ::File.exist?(license_path)
+              ::File.open(license_path, ::File::RDONLY) do |license_file|
+                license = YAML.load(license_file.read)
+                if product.license.name == license['license_name']
```
Comment (tyler-ball) [inline components/ruby/lib/license_acceptance/strategy/file.rb:36]: If users have previously accepted a license for (IE) Chef Infra, we do not need to re-prompt them for acceptance, even if their file does not include the `license_name`.

I think we need to assume they accepted `'EULA'` if `license_name` is not present.

Comment (clintoncwolfe) [reply components/ruby/lib/license_acceptance/strategy/file.rb]: 👍 Implemented in 4b87615

File: components/ruby/config/product_info.toml (hunk 1 above)
Comment (clintoncwolfe) [reply components/ruby/config/product_info.toml:1]: In fact, it is still a true statement - this is a misleading changeset. There is an expeditor [configuration](https://github.com/chef/license-acceptance/blob/06c48a3baff7ba37de29164fb8bb88e49deb53c0/.expeditor/config.yml#L41) that copies /product_info.toml to components/ruby/config/product_info.toml. The [script](https://github.com/chef/license-acceptance/blob/master/.expeditor/copy_product_info.sh) also inserts the warning line. What has happened here is that Tyler has manually copied the file to sync it, and it thus does not have the warning line.

**Review by marcparadise — CHANGES_REQUESTED** [review]
One minor request inline

File: components/ruby/lib/license_acceptance/product_builder.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+require_relative("product")
+require_relative("license_list")
+
+module LicenseAcceptance
+  class ProductBuilder
+    def self.build
+      builder = new
+      yield(builder)
+      builder.product
+    end
```
Comment (phiggins) [inline components/ruby/lib/license_acceptance/product_builder.rb:12]: Any reason not to put this method on Product and call use the accessors as the interface?

File: components/ruby/lib/license_acceptance/strategy/file.rb [hunk 5]
```diff
@@ -29,11 +29,23 @@ def accepted?(product_relationship)
 
         searching.each do |product|
           config.license_locations.each do |loc|
-            f = ::File.join(loc, product.filename)
-            if ::File.exist?(f)
-              logger.debug("Found license #{product.filename} at #{f}")
-              missing_licenses.delete(product)
-              break
+            license_path = ::File.join(loc, product.filename)
+            if ::File.exist?(license_path)
```
Comment (phiggins) [inline components/ruby/lib/license_acceptance/strategy/file.rb:33]: You could avoid a level of nesting by returning early from the block.

```suggestion
            next unless ::File.exist?(license_path)
```

File: components/ruby/lib/license_acceptance/strategy/file.rb [hunk 6]
```diff
@@ -29,11 +29,23 @@ def accepted?(product_relationship)
...
+                license["license_name"] ||= "EULA"
+
+                if product.license.name == license["license_name"]
+                  # It is possible to have previously accepted a product under the EULA license but that
+                  # product is actually covered under MLSA, so we need to check it.
+                  logger.debug("Found license #{product.filename} at #{license_path}")
+                  missing_licenses.delete(product)
+                  break
+                end
+              end
```
Comment (phiggins) [inline components/ruby/lib/license_acceptance/strategy/file.rb:48]: `YAML.load_file` could be used here to avoid a level of nesting and some file boilerplate.

```suggestion
              license = YAML.load_file(license_path)

              # If the license file is missing the license_name field,
              # assume it represents the EULA
              license["license_name"] ||= "EULA"

              if product.license.name == license["license_name"]
                # It is possible to have previously accepted a product under the EULA license but that
                # product is actually covered under MLSA, so we need to check it.
                logger.debug("Found license #{product.filename} at #{license_path}")
                missing_licenses.delete(product)
                break
              end
```

File: components/ruby/lib/license_acceptance/strategy/file.rb [hunk 7]
```diff
@@ -29,11 +29,23 @@ def accepted?(product_relationship)
...
+            license_path = ::File.join(loc, product.filename)
+            if ::File.exist?(license_path)
+              ::File.open(license_path, ::File::RDONLY) do |license_file|
+                license = YAML.load(license_file.read)
+
+                # If the license file is missing the license_name field,
+                # assume it represents the EULA
+                license["license_name"] ||= "EULA"
+
+                if product.license.name == license["license_name"]
```
Comment (phiggins) [inline components/ruby/lib/license_acceptance/strategy/file.rb:41]: `product.license_name` might be a convenient delegation to use here.

**Comment by marcparadise** [comment]
Converted this to draft until we discuss customer impacts of license changes

File: components/ruby/lib/license_acceptance/product_builder.rb (hunk 4 above)
Comment (clintoncwolfe) [reply components/ruby/lib/license_acceptance/product_builder.rb:12]: Well, the Builder design pattern has them as separate classes. I'm not sure Builder is really needed here, though - I think just switching to a constructor with keyword args would be simpler and would involve less overhead in calling it.

**Comment by btm** [comment]
> @tas50: Maybe this isn't the place for this, but WTF. Why don't we just use the new EULA on everything.

This isn't this place for my full reply, mostly because IANAL (but I do talk to ours). First, the list in the issue isn't final, we're reworking it with legal. So don't worry about the list yet. But to your question, there are a couple reasons we don't use the EULA on everything.

One is that some of our old proprietary closed source software such as Chef Manage is still only available for use with a commercial agreement. The EULA is explicitly not a commercial agreement, but the MLSA is. At this time we have no plans to open source Chef Manage before or after it becomes end of life.

Another is that, maybe, technically, some software that we don't explicitly sell like Chef Workstation wouldn't by default have a commercial agreement to replace the EULA for customers as it wouldn't be listed by name on an Order Form.
